tempesta_fw/http_parser.c

```diff
@@ -1323,28 +1342,70 @@ tfw_http_parse_req(void *req_data, unsigned char *data, size_t len)
 	 *
 	 * We must not rewrite abs_path, but still can cast host part
 	 * to lower case.
+	 *
+	 * We parse authority. It can be "host" or "userinfo@host".
+	 * At the begining we don't know, which of variants we have.
+	 *
+	 * So we start to fill both req->host and req->userinfo. If we find
+	 * '@' we reset req->host and finish req->userinfo. If we don't,
+	 * we clean req->userinfo.
 	 */
-	__FSM_STATE(Req_UriHostStart) {
+	__FSM_STATE(Req_UriAuthorityStart) {
 		if (likely(isalnum(c) || c == '.' || c == '-')) {
 			*p = LC(*p);
+			tfw_http_msg_set_data(msg, &req->userinfo, p);
 			tfw_http_msg_set_data(msg, &req->host, p);
-			__FSM_MOVE_f(Req_UriHost, &req->host);
+			__FSM_MOVE_f2(Req_UriAuthority, &req->host, &req->userinfo);
 		}
 		return TFW_BLOCK;
 	}
 
-	__FSM_STATE(Req_UriHost) {
-		if (likely(isalnum(c) || c == '.' || c == '-')) {
+	__FSM_STATE(Req_UriAuthority) {
+		if (likely(isalnum(c) || c == '.' || c == '-' ||
+			(c == '@' && !(parser->flags & TFW_HTTP_PF_HAS_USERINFO))))
+		{
 			*p = LC(*p);
-			__FSM_MOVE_f(Req_UriHost, &req->host);
+
+			if (c == '@') {
+				__field_finish(msg, &req->userinfo, data, p);
+				parser->flags |= TFW_HTTP_PF_HAS_USERINFO;
+				/* We need to reset req->host.
+				 * But now we on the '@' symbol.
+				 * So we should reset it on the next character */
+				parser->flags |= TFW_HTTP_PF_HOST_RESET;
+				__FSM_MOVE_f(Req_UriAuthority, &req->host);
+			}
+
+			if (parser->flags & TFW_HTTP_PF_HOST_RESET) {
+				tfw_http_msg_set_data(msg, &req->host, p);
+				parser->flags &= ~TFW_HTTP_PF_HOST_RESET;
+			}
+
+			if (parser->flags & TFW_HTTP_PF_HAS_USERINFO) {
+				__FSM_MOVE_f(Req_UriAuthority, &req->host);
+			} else {
+				__FSM_MOVE_f2(Req_UriAuthority,
+					      &req->host, &req->userinfo);
+			}
```

Comment:
Too much work is done in the sate for userinfo which is quite rare. You should not write ```req->host``` and ```req->userinfo``` at the same time. Rather, do the same things as previously - just write to ```req->host```. However, if you face '@' (you need to add it to allowed alphabet for the state), then just copy ```req->host``` to ```req->userinfo``` if ```TFW_STR_EMPTY(&req->userinfo). Otherwise (if we already added TFW_STR_COMPLETE flag to ```req->userinfo.flags```) block the request. Next, at '@' initialize ```req->host```, and you can continue to process the field in the same state - now you're writing to ```req->host```.